src/django_components/dependencies.py

```diff
@@ -0,0 +1,764 @@
+"""All code related to management of component dependencies (JS and CSS scripts)"""
+
+import json
+import re
+import sys
+from abc import ABC, abstractmethod
+from functools import lru_cache
+from hashlib import md5
+from typing import (
+    TYPE_CHECKING,
+    Callable,
+    Dict,
+    Iterable,
+    List,
+    Literal,
+    NamedTuple,
+    Optional,
+    Set,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+)
+from weakref import WeakValueDictionary
+
+from asgiref.sync import iscoroutinefunction, markcoroutinefunction
+from django.forms import Media
+from django.http import HttpRequest, HttpResponse, HttpResponseNotAllowed, HttpResponseNotFound, StreamingHttpResponse
+from django.http.response import HttpResponseBase
+from django.templatetags.static import static
+from django.urls import path, reverse
+from django.utils.decorators import sync_and_async_middleware
+from django.utils.safestring import SafeString, mark_safe
+from selectolax.lexbor import LexborNode
+
+import django_components.types as types
+from django_components.html import insert_before_end, parse_node, transform_html_document
+from django_components.utils import escape_js_string_literal, get_import_path
+
+if TYPE_CHECKING:
+    from django_components.component import Component
+
+
+ScriptType = Literal["css", "js"]
+RenderKind = Literal["document", "fragment"]
+
+
+#########################################################
+# 1. Cache the inlined component JS and CSS scripts,
+#    so they can be referenced and retrieved later via
+#    an ID.
+#########################################################
+
+
+class ComponentMediaCacheABC(ABC):
+    @abstractmethod
+    def get(self, key: str) -> Optional[str]: ...  # noqa: #704
+
+    @abstractmethod
+    def has(self, key: str) -> bool: ...  # noqa: #704
+
+    @abstractmethod
+    def set(self, key: str, value: str) -> None: ...  # noqa: #704
+
+
+class InMemoryComponentMediaCache(ComponentMediaCacheABC):
+    def __init__(self) -> None:
+        self._data: Dict[str, str] = {}
+
+    def get(self, key: str) -> Optional[str]:
+        return self._data.get(key, None)
+
+    def has(self, key: str) -> bool:
+        return key in self._data
+
+    def set(self, key: str, value: str) -> None:
+        self._data[key] = value
+
+
+comp_media_cache = InMemoryComponentMediaCache()
```

Comment:
Initially I used Django's cache, and only then I refactored it to our own class. That's why I also defined it as an abstract class, so if there is ever a need to support different implementations for the cache (e.g. that users could supply their own), then we know that this is where to do it.